File: algorithms/264.Ugly_Number_II/Ugly_Number_II.py
<<<<<<< HEAD
class Solution(object):
    def nthUglyNumber(self, n):
        """
        :type n: int
        :rtype: int
        """
        ans=1
        q2,q3,q5=[[i] for i in [2,3,5]]
        while n>1:
            n-=1
            minval=min(q2[0],q3[0],q5[0])
            if minval==q2[0]:
                ans=q2.pop(0)
                q2.append(ans*2)
                q3.append(ans*3)
                q5.append(ans*5)
            elif minval==q3[0]:
                ans=q3.pop(0)
                q3.append(ans*3)
                q5.append(ans*5)
            else:
                ans=q5.pop(0)
                q5.append(ans*5)
        return ans

=======
class Solution(object):
    def nthUglyNumber(self, n):
        """
        :type n: int
        :rtype: int
        """
        ans=1
        q2,q3,q5=[[i] for i in [2,3,5]]
        while n>1:
            n-=1
            minval=min(q2[0],q3[0],q5[0])
            if minval==q2[0]:
                ans=q2.pop(0)
                q2.append(ans*2)
                q3.append(ans*3)
                q5.append(ans*5)
            elif minval==q3[0]:
                ans=q3.pop(0)
                q3.append(ans*3)
                q5.append(ans*5)
            else:
                ans=q5.pop(0)
                q5.append(ans*5)
        return ans

>>>>>>> 6200c8704614e918c8bfa5357c648dd1b4f7eb74

File: algorithms/264.Ugly_Number_II/Ugly_Number_II.cc
/*
We have an array k of first n ugly number. We only know, at the beginning, the first one, which is 1. Then

k[1] = min( k[0]x2, k[0]x3, k[0]x5). The answer is k[0]x2. So we move 2's pointer to 1. Then we test:

k[2] = min( k[1]x2, k[0]x3, k[0]x5). And so on. Be careful about the cases such as 6, in which we need to forward both pointers of 2 and 3.

x here is multiplication.
*/
class Solution {
public:
    int nthUglyNumber(int n) {
        if(n <= 0) return false; // get rid of corner cases 
        if(n == 1) return true; // base case
        int t2 = 0, t3 = 0, t5 = 0; //pointers for 2, 3, 5
        vector<int> k(n);
        k[0] = 1;
        for(int i  = 1; i < n ; i ++)
        {
            k[i] = min(k[t2]*2,min(k[t3]*3,k[t5]*5));
            if(k[i] == k[t2]*2) t2++; 
            if(k[i] == k[t3]*3) t3++;
            if(k[i] == k[t5]*5) t5++;
        }
        return k[n-1];
    }
};